File: frontend/src/features/twin/types/twinSimulation.ts
export interface PackInfo {
  slug: string;
  name: string;
  domain_count: number;
}

export interface PackState {
  enabled_globally: boolean;
}

export interface RecentHitSample {
  root_domain: string;
  device_id: number;
  hostname: string | null;
  last_seen_at: string;
  query_count_estimate: number;
}

export interface SimulationSummary {
  devices_affected: number;
  newly_blocked_domain_count: number;
  recent_hits_count: number;
  recent_hits_sample: RecentHitSample[];
}

export interface DeviceSimulationImpact {
  device_id: number;
  hostname: string | null;
  added_block_count: number;
  recent_hits: string[];
}

export interface PackToggleSimulationResponse {
  generated_at: string;
  lookback_hours: number;
  pack: PackInfo;
  current_state: PackState;
  proposed_state: PackState;
  summary: SimulationSummary;
  devices: DeviceSimulationImpact[];
}

export interface PackToggleSimulationRequest {
  pack_slug: string;
  enabled_globally: boolean;
}

File: backend/app/features/twin/schemas/twin_simulation.py
from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field


class PackToggleSimulationRequest(BaseModel):
    pack_slug: str
    enabled_globally: bool


class PackInfoRead(BaseModel):
    slug: str
    name: str
    domain_count: int


class PackStateRead(BaseModel):
    enabled_globally: bool


class RecentHitSample(BaseModel):
    root_domain: str
    device_id: int
    hostname: Optional[str] = None
    last_seen_at: datetime
    query_count_estimate: int = 1


class SimulationSummaryRead(BaseModel):
    devices_affected: int
    newly_blocked_domain_count: int
    recent_hits_count: int
    recent_hits_sample: List[RecentHitSample] = Field(default_factory=list)


class DeviceSimulationImpactRead(BaseModel):
    device_id: int
    hostname: Optional[str] = None
    added_block_count: int
    recent_hits: List[str] = Field(default_factory=list)


class PackToggleSimulationResponse(BaseModel):
    generated_at: datetime
    lookback_hours: int
    pack: PackInfoRead
    current_state: PackStateRead
    proposed_state: PackStateRead
    summary: SimulationSummaryRead
    devices: List[DeviceSimulationImpactRead] = Field(default_factory=list)
